### ISSUE TITLE
Changes in route matching in 0.23.7 -> 0.23.8

### ISSUE THREAD

**Issue by eikek** [http4s/http4s#5959]
Hi all,

I noticed some issues after upgrading to 0.23.8. Looking into it, it seems that request matching changed slightly between versions. Here is a small example:

```scala
#!/usr/bin/env amm

import $ivy.`org.http4s::http4s-dsl:0.23.7`
import $ivy.`org.http4s::http4s-blaze-server:0.23.7`
import $ivy.`ch.qos.logback:logback-classic:1.2.3`

import scala.concurrent.ExecutionContext
import cats.effect._
import fs2.Stream
import org.http4s._
import org.http4s.dsl.Http4sDsl
import org.http4s.server.Router
import org.http4s.blaze.server.BlazeServerBuilder
import org.http4s.implicits._

object TestApp extends IOApp.Simple {
  def routes: HttpRoutes[IO] = {
    val dsl = new Http4sDsl[IO] {}
    import dsl._

    HttpRoutes.of {
      case GET -> Root / id =>
        Ok(s"GET id: $id")

      case GET -> Root =>
        Ok("GET /")
    }
  }

  def app =
    Router("tasks" -> routes).orNotFound

  def server =
    BlazeServerBuilder[IO](ExecutionContext.global)
      .bindHttp(8888, "0.0.0.0")
      .withHttpApp(app)
      .serve

  val run =
    server.compile.drain
}

@main
def main(): Unit =
  TestApp.main(Array.empty)
```

When requesting `http://localhost:888/tasks`, it will match different routes:

```
# 0.23.7
❯ curl http://localhost:8888/tasks
GET /⏎

# 0.23.8
❯ curl http://localhost:8888/tasks
GET id: ⏎
```

With http4s 0.23.7 it matches the second case and with 0.23.8 it matches the first where `id` is empty (it's the same when appending a `/`).  Is this an expected change?

Thank you!

**Comment by armanbilge** [comment]
Thanks for the report. I haven't had a chance to look into whether this is an expected change or not, but it's almost definitely due to:
- https://github.com/http4s/http4s/pull/5794

**Comment by ChristopherDavenport** [comment]
I would expect that to no match. I think we expect segments there to be nonEmpty previously.

**Comment by hamnis** [comment]
This is indeed a problem caused in #5794 
A closer look at that is needed.

**Comment by rossabaker** [comment]
Fixed by #5991, will merge forward from 0.22.
